### PR TITLE
Fixed result buffer overflow in intersectConvexConvex_ for non-convex input

### LINKED ISSUE #25259: crash when invoke cv2.intersectConvexConvex
### System Information

Python version: 3.8.18
OpenCV python version: 4.9.0
Operating System / Platform: Windows-10-10.0.19045-SP0

### Detailed description

Process finished with exit code -1073740791 (0xC0000409)

### Steps to reproduce

```

import cv2
import numpy as np
p1 =np.array([[[ 46.077175 , 228.66121  ]],

       [[  5.428622 , 250.05899  ]],

       [[207.51741  , 109.645676 ]],

       [[175.94789  ,  32.6566   ]],

       [[217.4915   , 252.66176  ]],

       [[187.09386  ,   6.3988557]],

       [[ 52.20488  ,  69.266205 ]],

       [[ 38.188286 , 134.48068  ]],

       [[246.4742   ,  31.41043  ]],

       [[178.97946  , 169.52287  ]],

       [[103.40764  , 153.30397  ]],

       [[160.67746  ,  17.166115 ]],

       [[152.44255  , 135.35     ]],

       [[197.03804  , 193.04782  ]],

       [[248.28397  ,  56.821487 ]],

       [[ 10.907227 ,  82.55291  ]],

       [[109.67949  ,  70.7405   ]],

       [[ 58.96842  , 150.132    ]],

       [[150.7613   , 129.54753  ]],

       [[254.98463  , 228.21748  ]],

       [[139.02563  , 193.89336  ]],

       [[ 84.79946  , 162.25363  ]],

       [[ 39.83567  ,  44.626484 ]],

       [[107.034996 , 209.38887  ]],

       [[ 67.61073  ,  17.119232 ]],

       [[208.8617   ,  33.67367  ]],

       [[182.65207  ,   8.291072 ]],

       [[ 72.89319  ,  42.51845  ]],

       [[202.4902   , 123.97209  ]],

       [[ 79.945076 , 140.99268  ]],

       [[225.8952   ,  66.226326 ]],

       [[ 34.08404  , 219.2208   ]],

       [[243.1221   ,  60.95162  ]]], dtype=np.float32)

p2 = np.array([[[144.33624  , 247.15732  ]],

       [[  5.656847 ,  17.461054 ]],

       [[230.54338  ,   2.0446582]],

       [[143.0578   , 215.27856  ]],

       [[250.44626  ,  82.54287  ]],

       [[  0.3846766,  11.101262 ]],

       [[ 70.81022  ,  17.243904 ]],

       [[ 77.18812  ,  75.760666 ]],

       [[190.34933  , 234.30962  ]],

       [[230.10204  , 133.67998  ]],

       [[ 58.903755 , 252.96451  ]],

       [[213.57228  , 155.7058   ]],

       [[190.80992  , 212.90802  ]],

       [[203.4356   ,  36.55016  ]],

       [[ 32.276424 ,   2.5646307]],

       [[ 39.73823  ,  87.23782  ]],

       [[112.46902  , 101.81753  ]],

       [[ 58.154305 , 238.40395  ]],

       [[187.01064  ,  96.24343  ]],

       [[ 44.42692  ,  10.573529 ]],

       [[118.76949  , 233.35114  ]],

       [[ 86.26109  , 120.93148  ]],

       [[217.94751  , 130.5933   ]],

       [[148.2687   ,  68.56015  ]],

       [[187.44174  , 214.32857  ]],

       [[247.19875  , 180.8494   ]],

       [[ 17.986013 ,  61.451443 ]],

       [[254.74344  , 204.71747  ]],

       [[211.92726  , 132.0139   ]],

       [[ 51.36624  , 116.63085  ]],

       [[ 83.80044  , 124.20074  ]],

       [[122.125854 ,  25.182402 ]],

       [[ 39.08164  , 180.08517  ]]], dtype=np.float32)

out1, out2 = cv2.intersectConvexConvex(p1, p2, False)



### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

### PR DESCRIPTION
**Pull request by asmorkalov**
Fixes #25259

The reproducer in the ticket is not valid. It uses non-convex inputs. The PR fixes uncontrolled buffer overflow to prevent crashes.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by asmorkalov** [comment]
@mshabunin @opencv-alalek any comment it this?

File: modules/imgproc/test/test_intersectconvexconvex.cpp [hunk 1]
```diff
@@ -255,6 +255,42 @@ TEST(Imgproc_IntersectConvexConvex, intersection_4)
...
+        {187.01064f  ,  96.24343f  }, { 44.42692f  ,  10.573529f }, {118.76949f  , 233.35114f  },
+        { 86.26109f  , 120.93148f  }, {217.94751f  , 130.5933f   }, {148.2687f   ,  68.56015f  },
+        {187.44174f  , 214.32857f  }, {247.19875f  , 180.8494f   }, { 17.986013f ,  61.451443f },
+        {254.74344f  , 204.71747f  }, {211.92726f  , 132.0139f   }, { 51.36624f  , 116.63085f  },
+        { 83.80044f  , 124.20074f  }, {122.125854f ,  25.182402f }, { 39.08164f  , 180.08517f  }
+    };
+    std::vector<cv::Point> intersection;
+
+    float area = cv::intersectConvexConvex(convex1, convex2, intersection, false);
+    EXPECT_TRUE(intersection.empty());
```
Comment (mshabunin) [inline modules/imgproc/test/test_intersectconvexconvex.cpp:291]: What if `intersection` was non-empty before function call, will it be empty afterwards? Maybe we could check it here too.

**Review by mshabunin — APPROVED** [review]
Looks good to me! Just one minor comment/suggestion for the test.

File: modules/imgproc/test/test_intersectconvexconvex.cpp (hunk 1 above)
Comment (asmorkalov) [reply modules/imgproc/test/test_intersectconvexconvex.cpp:291]: The function calls `cv::OutputArray::release()` for the "error" case, so it's always empty after the test call.

**Pull request merged**